sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/spi/HttpClientProviders.java

```diff
@@ -0,0 +1,38 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.core.implementation.http.spi;
+
+import com.azure.core.http.HttpClient;
+
+import java.util.Iterator;
+import java.util.ServiceLoader;
+
+/**
+ * This class handles loading available HTTP clients
+ */
+public final class HttpClientProviders {
+    private static HttpClientProvider defaultProvider;
+
+    static {
+        ServiceLoader<HttpClientProvider> serviceLoader = ServiceLoader.load(HttpClientProvider.class);
+
+        // Use the first provider found in the service loader iterator.
+        Iterator<HttpClientProvider> it = serviceLoader.iterator();
```

Comment:
This allows for multiple implementations of `HttpClientProvider` to be available in the classpath. Also, as per [documentation of iterator](https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html#iterator--), implementations are returned in instantiation order which can be different each time the application loads and hence result in the unpredictability of the `HttpClientProvider` used. 

Given that this is critical for network communication, we should try and make this more predictable:
1. Don't allow more than one implementation to be available in the classpath. Throw exception if more than one found.
2. Sort the available implementations in alphabetical order and pick the first - at least the same implementation is picked each time the application runs.

Thoughts?

Reply:
I believe that instantiation order is decided by classpath order. If this is correct, given no changes to the classpath, the order should always be the same.

On the one hand I agree with your concerns. On the other, there might be valid use cases for multiple http clients, when our builder API supports a user specifying a specific http client in a particular case. I would agree that this is a vanishingly small use case however.